jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java

```diff
@@ -0,0 +1,561 @@
+            {
+                return dstPort;
+            }
+
+            @Override
+            public ClientConnectionFactory apply(ClientConnectionFactory factory)
+            {
+                return new V1(factory);
+            }
+
+            @Override
+            public boolean equals(Object obj)
+            {
+                if (this == obj)
+                    return true;
+                if (obj == null || getClass() != obj.getClass())
+                    return false;
+                Tag that = (Tag)obj;
+                return Objects.equals(family, that.family) &&
+                    Objects.equals(srcIP, that.srcIP) &&
+                    srcPort == that.srcPort &&
+                    Objects.equals(dstIP, that.dstIP) &&
+                    dstPort == that.dstPort;
+            }
+
+            @Override
+            public int hashCode()
+            {
+                return Objects.hash(family, srcIP, srcPort, dstIP, dstPort);
+            }
+        }
+    }
+
+    public static class V2 extends ProxyProtocolClientConnectionFactory
+    {
+        public V2(ClientConnectionFactory factory)
+        {
+            super(factory);
+        }
+
+        @Override
+        protected ProxyProtocolConnection newProxyProtocolConnection(EndPoint endPoint, Map<String, Object> context)
+        {
+            HttpDestination destination = (HttpDestination)context.get(HttpClientTransport.HTTP_DESTINATION_CONTEXT_KEY);
+            Executor executor = destination.getHttpClient().getExecutor();
+            Tag tag = (Tag)destination.getOrigin().getTag();
+            if (tag == null)
+            {
+                InetSocketAddress local = endPoint.getLocalAddress();
+                InetSocketAddress remote = endPoint.getRemoteAddress();
+                boolean ipv4 = local.getAddress() instanceof Inet4Address;
+                tag = new Tag(Tag.Command.PROXY, ipv4 ? Tag.Family.INET4 : Tag.Family.INET6, Tag.Protocol.STREAM, local.getAddress().getHostAddress(), local.getPort(), remote.getAddress().getHostAddress(), remote.getPort());
+            }
+            return new ProxyProtocolConnectionV2(endPoint, executor, getClientConnectionFactory(), context, tag);
+        }
+
+        public static class Tag implements ClientConnectionFactory.Decorator
+        {
+            /**
+             * The PROXY V2 Tag typically used to "ping" the server.
+             */
+            public static final Tag LOCAL = new Tag(Command.LOCAL, Family.UNSPEC, Protocol.UNSPEC, null, 0, null, 0);
+
+            private Command command;
+            private Family family;
+            private Protocol protocol;
+            private String srcIP;
+            private int srcPort;
+            private String dstIP;
+            private int dstPort;
+            private Map<Integer, byte[]> vectors;
+
+            public Tag()
+            {
+                this(null, 0);
+            }
+
+            public Tag(String srcIP, int srcPort)
```

Comment:
probably should javadoc these constructors to say that null values mean that the proxy protocol will take the value from the EndPoint when then tag is applied to a specific connection (and not defaulted at construction).

Reply:
Ditto v1